### PR TITLE
Fixed metadata generator not producing metadata when the entry assembly lacked functions but dependent assemblies had them

### LINKED ISSUE #2294: No functions are recognized after upgrading Worker.Sdk to 1.17.0
### Description

After upgrading `Microsoft.Azure.Functions.Worker.Sdk` to **1.17.0**, the functions on _Class Library_ projects aren't not recognized/founded. It happens just if not exists functions on the Azure Function project.

```
Starting JobHost
Starting Host (HostId=apsnt1085-85705262, InstanceId=c9904c45-f99b-4bc5-8e51-58cc8b022ef6, Version=4.28.5.21962, ProcessId=27588, AppDomainId=1, InDebugMode=False, InDiagnosticMode=False, FunctionsExtensionVersion=(null))
Loading functions metadata
Worker indexing is enabled
Fetching metadata for workerRuntime: dotnet-isolated
Reading functions metadata (Worker)
0 functions found (Worker)
Reading functions metadata (Custom)
1 functions found (Custom)
0 functions loaded
Generating 0 job function(s)
No job functions found. Try making your job classes and methods public. If you're using binding extensions (e.g. Azure Storage, ServiceBus, Timers, etc.) make sure you've called the registration method for the extension(s) in your startup code (e.g. builder.AddAzureStorage(), builder.AddServiceBus(), builder.AddTimers(), etc.).
```

### Steps to reproduce

1. Create "Class Library" project
2. Add `Microsoft.Azure.Functions.Worker.Extensions.ServiceBus` package reference
3. Create a ServiceBus Function using `ServiceBusTrigger`
4. Create Azure Functions project
5. Update package `Microsoft.Azure.Functions.Worker.Sdk` to **1.17.0**
6. Delete `Function1.cs` file
7. Add reference of "Class Library" created on step 1

### PR DESCRIPTION
**Pull request by kshyju**
Fixed metadata generator not producing metadata when the entry assembly lacked functions but dependent assemblies had them

resolves #2294 

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

**Review by satvu — APPROVED** [review]
lgtm!

**Comment by kshyju** [comment]
Pushed an iteration after rebasing.

File: sdk/Sdk.Generators/Extensions/ListExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+namespace Microsoft.Azure.Functions.Worker.Sdk.Generators
+{
+    internal static class ListExtensions
+    {
+        /// <summary>
+        /// If the enumerable passed in is already a list, it is returned, otherwise ToList() is invoked.
+        /// </summary>
+        /// <typeparam name="T">The type of element in the list.</typeparam>
+        /// <param name="source">The enumerable to return as a list.</param>
+        internal static List<T> AsList<T>(this IEnumerable<T>? source)
```
Comment (fabiocav) [inline sdk/Sdk.Generators/Extensions/ListExtensions.cs:16]: Can you expand on why you feel this extension is needed in this scenario over simply using `ToList`?

The default behavior will trigger an allocation, but this is only being avoided for one of the methods you have currently using this. Is there something specific you're trying to address with this? Did you observe significant issues/impact when using `ToList`?

Comment (kshyju) [reply sdk/Sdk.Generators/Extensions/ListExtensions.cs]: Synced offline. Pushed an update - Removed the extension method.

**Review by fabiocav — APPROVED** [review]

**Pull request merged**